db/snapshot/backup_task.cc

```diff
@@ -113,6 +114,10 @@ future<> backup_task_impl::do_backup() {
     }
 
     std::exception_ptr ex;
+    // FIXME: this should be configurable, or derived dynamically from
+    // the concurrency level of S3 uploads
+    constexpr size_t max_concurrency = sstables::num_component_types * 2;
```

Comment:
The concurrency should be in terms of bytes in flight, not components. Lots of small components (could be small sstables) should be backed up concurrently, while one large component should have a concurrency of 1 to get rid of it quickly. 

If we guess 250ms latency to seal a file, and want 100 MB/s throughput/shard, we need 25MB worth of components in flight.

Reply:
do we put each sstable in its own s3 prefix to improve throughput (this might be important especially for tiered storage)? don't we have to limit overall number of in-flight components to limit mem usage or we'll have that as a secondary effect of targeting a certain throughput?

Reply:
@xemul ?

Reply:
> do we put each sstable in its own s3 prefix to improve throughput (this might be important especially for tiered storage)? 

No we don't. Note, there are two distinct usages of s3 client and, respectively, the prefixes and names that are used by sstables components   -- keyspace storage options and backup/restore

> don't we have to limit overall number of in-flight components to limit mem usage or we'll have that as a secondary effect of targeting a certain throughput?

There's both -- mem semaphore for S3 buffers and maximum number of connections an underlying http client mayhave

Reply:
In v3 ([7ab3391](https://github.com/scylladb/scylladb/commit/7ab3391f15e68adfff6187385be438b96bb3d57e)): using the sstable dir semaphore